api/dimensions.go

```diff
@@ -108,34 +74,13 @@ func (api *DatasetAPI) getDimensions(w http.ResponseWriter, r *http.Request) {
 			slicedResults = utils.Slice(results, offset, limit)
 		}
 
-		listOfDimensions := &models.DatasetDimensionResults{
-			Items:      slicedResults,
-			Offset:     offset,
-			Limit:      limit,
-			Count:      len(slicedResults),
-			TotalCount: len(dimensions),
-		}
-
-		b, err := json.Marshal(listOfDimensions)
-		if err != nil {
-			log.Event(ctx, "failed to marshal list of dimension resources into bytes", log.ERROR, log.Error(err), logData)
-			return nil, err
-		}
-		return b, nil
+		return slicedResults, len(dimensions), nil
 	}()
 	if err != nil {
 		handleDimensionsErr(ctx, w, "", err, logData)
-		return
+		return nil, 0, err
 	}
-
-	setJSONContentType(w)
-	_, err = w.Write(b)
-	if err != nil {
-		log.Event(ctx, "error writing bytes to response", log.ERROR, log.Error(err), logData)
-		http.Error(w, err.Error(), http.StatusInternalServerError)
-	}
-
-	log.Event(ctx, "getDimensions endpoint: request successful", log.INFO, logData)
+	return list, totalCount, nil
```

Comment:
If I am following this correctly this function returns the success values (the list and total count) to the paginator. The paginator is then responsible for marshalling and writing the response body. **However**, if there is an error then its this package that is responsible for writing an error response. Having the http logic split across 2 different packages depending on the outcome feels wrong to me. 


IMO the paginator is now fulfilling the role of `http.Handler` and this function should simply execute the business logic - it's no longer a `http.Handler` its a plain old function that given some inputs returns a list of the dimension options. 

The paginator should use this func to get the data. It should then be responsible for writing the data to a http response or writing an error response if it was unsuccessful. As it is the responsibility or writing the response is split between the 2. I think it would cleaner and clearer if there were a clear separation of concerns `http.Handler` and  `listGetterFunction`.

I appreciate this is a complex issue to solve and you are trying to work out a design that fits into the existing code and I'm not making it any easier with these comments 😂  _but_ my motivation comes from the right place. I suggest we have a call tomorrow and we can discuss and possible work out a solution together?

Reply:
I was thinking this might just be a case of moving the `handleDimensionsErr()` into the paginator? Maybe update the paginator to take an `handleErrorFunc()` param. That would allow you to customise how the errors from the list getter are mapped into a HTTP response without having to hardcode it.

Reply:
As we discussed this comment will be addressed in this ticket: https://trello.com/c/Ichr8HBF/429-refactor-in-place-the-existing-pagination-code-so-that-responses-are-only-written-in-one-place